lib/util/id3_utils.js

```diff
@@ -0,0 +1,298 @@
+    } else if (frame.type === 'WXXX') {
+      /*
+       * Format:
+       * [0]   = {Text Encoding}
+       * [1-?] = {Description}\0{URL}
+       */
+      if (frame.size < 2) {
+        return null;
+      }
+      if (frame.data[0] !== shaka.util.Id3Utils.UTF8_encoding) {
+        shaka.log.warning('Ignore frame with unrecognized character ' +
+            'encoding');
+        return null;
+      }
+      const descriptionEndIndex = frame.data.subarray(1).indexOf(0);
+
+      if (descriptionEndIndex === -1) {
+        return null;
+      }
+      const description = StringUtils.fromUTF8(
+          BufferUtils.toUint8(frame.data, 1, descriptionEndIndex));
+      const data = StringUtils.fromUTF8(
+          BufferUtils.toUint8(frame.data, description.length + 1));
+
+      metadataFrame.description = description;
+      metadataFrame.data = data;
+      metadataFrame.value = metadataFrame.data;
+      return metadataFrame;
+    } else if (frame.type === 'PRIV') {
+      /*
+       * Format: <text string>\0<binary data>
+       */
+      if (frame.size < 2) {
+        return null;
+      }
+      const ownerEndIndex = frame.data.indexOf(0);
+      if (ownerEndIndex === -1) {
+        return null;
+      }
+      const owner = StringUtils.fromUTF8(
+          BufferUtils.toUint8(frame.data, 0, ownerEndIndex));
+      const data = BufferUtils.toArrayBuffer(
+          frame.data.subarray(owner.length + 1));
+      metadataFrame.description = owner;
+      metadataFrame.data = data;
+      metadataFrame.value = metadataFrame.data;
+      return metadataFrame;
+    } else if (frame.type[0] === 'T') {
+      /*
+       * Format:
+       * [0]   = {Text Encoding}
+       * [1-?] = {Value}
+       */
+      if (frame.size < 2) {
+        return null;
+      }
+      if (frame.data[0] !== shaka.util.Id3Utils.UTF8_encoding) {
+        shaka.log.warning('Ignore frame with unrecognized character ' +
+            'encoding');
+        return null;
+      }
+      const text = StringUtils.fromUTF8(frame.data.subarray(1));
+      metadataFrame.data = text;
+      metadataFrame.value = metadataFrame.data;
+      return metadataFrame;
+    } else if (frame.type[0] === 'W') {
+      /*
+       * Format:
+       * [0-?] = {URL}
+       */
+      const url = StringUtils.fromUTF8(frame.data);
+      metadataFrame.data = url;
+      metadataFrame.value = metadataFrame.data;
+      return metadataFrame;
+    }
+
+    shaka.log.warning('Unrecognized/unsupported ID3 frame type:', frame.type);
+
+    return null;
+  }
+
+  /**
+   * Returns an array of ID3 frames found in all the ID3 tags in the id3Data
+   * @param {Uint8Array} id3Data - The ID3 data containing one or more ID3 tags
+   * @return {!Array.<shaka.extern.MetadataFrame>}
+   */
+  static getID3Frames(id3Data) {
+    const Id3Utils = shaka.util.Id3Utils;
+    let offset = 0;
+    const frames = [];
+
+    while (offset < (id3Data.length - 1000)) {
+      while (Id3Utils.isHeader_(id3Data, offset)) {
```

Comment:
I wonder if it would be useful to replace isHeader_ with something that actually parses and returns the header instead.  I think if you end up needing to deal with extended headers or unsynchronisation, that will probably be necessary.

Reply:
Right now I think it's not necessary, so we don't complicate the logic.